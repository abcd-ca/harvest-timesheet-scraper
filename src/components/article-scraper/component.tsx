import React, { FunctionComponent } from "react";
import "./styles.scss";
import { browser, Tabs } from "webextension-polyfill-ts";

// // // //

interface ArticleScraperState {
  article: string;
}

export class ArticleScraper extends React.Component<{}, ArticleScraperState> {
  constructor(props: Readonly<{}>) {
    super(props);
    this.state = { article: "..." };
  }

  scrapeArticle(): void {
    function getMediumArticle(): string {
      const title = document.querySelector(
        "#root > div > article > div > section > div > div > div > div > h1",
      )?.textContent;

      const subtitle = document.querySelector(
        "#root > div > article > div > section > div > div > div > h2",
      )?.textContent;

      const bodyNodes =
        document.querySelector(
          "#root > div > article > div > section > div > div",
        )?.children || [];

      const author = document.querySelector(
        "#root > div > article > div > section > div > div > div > div > div > div > div > div > span > div > span > a",
      )?.textContent;

      let content = "";

      for (let i = 0; i < bodyNodes.length; i++) {
        const node = bodyNodes[i]

        // Ignore divs
        if (["H1", "H2", "H3", "P", "BLOCKQUOTE"].indexOf(node.tagName) != -1) {
          // console.log(node.textContent)
          content += node.textContent + "\n\n"
        }
      }

      // console.log([title, subtitle, author, content])

      content = content.replace(/[“”]/g, '"')
                       .replace(/[‘’]/g, "'")
                       .replace(/\—/g, "-")

      return content
    }

    browser.tabs
      .query({ active: true, currentWindow: true })
      .then((tabs: Tabs.Tab[]) => {
        // Pulls current tab from browser.tabs.query response
        const currentTab: Tabs.Tab | undefined = tabs[0];

        // Short circuits function execution is current tab isn't found
        if (!currentTab) {
          return;
        }

        // browser.runtime.sendMessage({ popupMounted: true });
        // Executes the script in the current tab
        browser.tabs
          .executeScript(currentTab.id, {
            code: "(" + getMediumArticle + ")();",
          })
          .then(article => {
            this.setState({ article: `${article}` })
          });
      });
  }

  render(): JSX.Element {
    return (
      <div className="row">
        <div className="col-lg-12">
          <button
            className="btn btn-block btn-outline-dark"
            onClick={(): void => this.scrapeArticle()}
          >
            Scrape Medium.com Article
          </button>

          <textarea
            name="result"
            id="result"
            className="form-control"
            cols={30}
            rows={10}
            value={this.state.article}
            readOnly
          ></textarea>

          {/* <button className="btn btn-block btn-outline-dark">Copy</button> */}
        </div>
      </div>
    );
  }
}
