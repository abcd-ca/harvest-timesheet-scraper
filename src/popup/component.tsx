import React, { FunctionComponent } from "react";
import { Hello } from "@src/components/hello";
import { browser } from "webextension-polyfill-ts";
// import { Scroller } from "@src/components/scroller";
import { ArticleScraper } from "@src/components/article-scraper";

// // // //

export const Popup: FunctionComponent = () => {
  // Sends the `popupMounted` event
  React.useEffect(() => {
    browser.runtime.sendMessage({ popupMounted: true });
  }, []);

  // Renders the component tree
  return (
    <div className="popup-container">
      <div className="container mx-4 my-4">
        <Hello />
        <hr />
        <ArticleScraper />
      </div>
    </div>
  );
};
