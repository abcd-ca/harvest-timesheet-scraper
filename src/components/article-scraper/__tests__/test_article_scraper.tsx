import * as React from "react";
import { ArticleScraper } from "../component";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";

it("component renders", () => {
    const tree: ReactTestRendererJSON | null = renderer
        .create(<ArticleScraper />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
