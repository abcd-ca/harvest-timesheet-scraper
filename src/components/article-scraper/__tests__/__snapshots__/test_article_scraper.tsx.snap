// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component renders 1`] = `
<div
  className="row"
>
  <div
    className="col-lg-12"
  >
    <button
      className="btn btn-block btn-outline-dark"
      onClick={[Function]}
    >
      Scrape Medium.com Article
    </button>
    <textarea
      className="form-control"
      cols={30}
      id="result"
      name="result"
      readOnly={true}
      rows={10}
      value="..."
    />
  </div>
</div>
`;
